core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java

```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
+        assertEquals(1, seqs.size);
+        assertEquals(smallSeq, seqs.head.getDefinition());
+        final ParseValue c = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c.asNumeric().intValue());
+        assertEquals(2, c.getOffset());
+    }
+
+    @Test
+    public void getAllRootsMulti() throws IOException {
+        final Token topSeq = seq(any("a"), smallSeq, smallSeq);
+        final ParseResult result = topSeq.parse(stream(1, 2, 3, 2, 3), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, smallSeq);
+        assertEquals(2, seqs.size);
+        assertEquals(smallSeq, seqs.head.getDefinition());
+        assertEquals(smallSeq, seqs.tail.head.getDefinition());
+        final ParseValue c1 = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
+    private HashSet<ParseItem> makeSet(ParseItemList seqs) {
+        final HashSet<ParseItem> items = new HashSet<>();
+        for (ParseItemList current = seqs; current != null && !current.isEmpty(); current = current.tail) {
+            items.add(current.head);
+        }
+        return items;
+    }
+
+    @Test
+    public void getAllRootsMultiSub() throws IOException {
+        final ParseResult result = rep(seq(smallSeq, sub(smallSeq, currentOffset))).parse(stream(1, 2, 1, 2, 1, 2, 1, 2), enc());
+                                                                                           /* 1: +--------+
+                                                                                           /* 2:       +--------+
+                                                                                           /* 3:             +--------+ */
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, smallSeq);
+        assertEquals(6, seqs.size); // Three regular and three subs.
+        final HashSet<ParseItem> items = makeSet(seqs);
+        assertEquals(seqs.size, items.size()); // Check that there are no duplicate results.
+        for (ParseItem item : items) {
+            assertTrue(item.isGraph());
+            assertEquals(2, item.asGraph().size);
+            assertEquals(2, item.asGraph().head.asValue().asNumeric().intValue());
+        }
+    }
+
+    private class CustomToken extends Token {
+
+        public final Token token;
+
+        public CustomToken() {
+            super("", enc());
+            token = seq(any("a"), opt(this));
+        }
+
+        @Override
+        protected ParseResult parseImpl(String scope, Environment env, Encoding enc) throws IOException {
+            return token.parse(scope, env, enc);
+        }
+    }
```

Comment:
I think I now see your point. The `CustomToken` instance itself never ends up as a `definition` of any `ParseItem` in the graph, but the nested `token` does.